File: hw1/hw1.tests/ParserTests.cs
using Xunit;

namespace hw1.tests
{
    public class ParserTests
    {
        [Theory]
        [InlineData(new[] {"1", "+", "2"}, 0)]
        [InlineData(new[] {"4", "-", "1"}, 0)]
        [InlineData(new[] {"3", "+", "1"}, 0)]
        
        public void ParseRight_WhenArgumentsRight(string[] args, int expected)
        {
            var res = Parser.TryToParse(args, out var val1,
                out var operation, out var val2);

            Assert.Equal(expected, res);
        }
        
        [Theory]
        [InlineData(new[] {"Gh", "+", "3"}, 1)]
        [InlineData(new[] {"", "+", "a"}, 1)]
        [InlineData(new[] {"aaaaaaa", "+", "ffffff"}, 1)]
        [InlineData(new[] {"", "+", ""}, 1)]
        [InlineData(new[] {"3","/","0"}, 3)]
        public void TryToParse_WithWrongArguments(string[] args, int expected)
        {
            var actual = Parser.TryToParse(args, out var val1,
                out var operation, out var val2);
            
            Assert.Equal(expected, actual);
        }
        
        [Theory]
        [InlineData(new[] {"777", "", "777"}, 2)]
        [InlineData(new[] {"77", "ololo", "77"}, 2)]
        [InlineData(new[] {"1","$","1"},2)]
        [InlineData(new[] {"2"," ","3"},2)]
        public void TryToParse_WithWrongOperator(string[] args, int expected)
        {
            var res = Parser.TryToParse(args, out var val1,
                out var operation, out var val2);
            
            Assert.Equal(expected, res);
        }
    }
}

File: hw5/hw5.Tests/ParserTests.fs
module ParserTests

open hw5.Parser
open hw5.Calculator
open Xunit


[<Theory>]
[<InlineData("1", "+", "2", 1, "+", 2 )>]
[<InlineData("2", "-", "1.1", 2, "-", 1.1 )>]
[<InlineData("2", "*", "0.002", 2, "*", 0.002 )>]
[<InlineData("2", "/", "5", 2, "/", 5 )>]
let ``ParseValidArgsAndOperations``(val1, operation, val2, expectArg1, expectOp, expectArg2) =
    let arg1, op, arg2 = TryParseArgs [|val1; operation; val2|]
    Assert.Equal(arg1, Some expectArg1)
    Assert.Equal(op, Some expectOp)
    Assert.Equal(arg2, Some expectArg2)

[<Theory>]
[<InlineData("2", "aaa", "3", 2, "aaa", 3 )>]
[<InlineData("2", "", "1.1", 2, "", 1.1 )>]
let ``ParseInvalidOperations``(val1, operation, val2, expectArg1, expectOp, expectArg2) =
    let arg1, op, arg2 = TryParseArgs [|val1; operation; val2|]
    Assert.Equal(op, None "NotSupportedOperation")
    
[<Theory>]
[<InlineData("aa", "+", "3", "?", "+", 3 )>]
[<InlineData("", "-", "3", "?", "-", 3 )>]
let ``ParseInvalidArgs``(val1, operation, val2, expectArg1, expectOp, expectArg2) =
    let arg1, op, arg2 = TryParseArgs [|val1; operation; val2|]
    Assert.Equal(arg1, None "InvalidArgument")
